### PR TITLE
Use params instead of env vars in *_install() tasks

### PR DESCRIPTION
**Pull request by lpramuk**
- install task params are added accordingly to the task type
- turn env vars into params in *_install() tasks
- cleanup

### CODE REVIEW

File: automation_tools/__init__.py [hunk 1]
```diff
@@ -1797,11 +1792,9 @@ def apply_hotfix():
         run('/root/hotfix.sh')
 
 
-def upstream_install(admin_password=None, run_katello_installer=True):
-    """Task to install Foreman nightly using forklift scripts"""
-    koji = 'koji' in os.environ.get('DISTRIBUTION', '').lower()
-    if admin_password is None:
-        admin_password = os.environ.get('ADMIN_PASSWORD', 'changeme')
+def upstream_install(admin_password='changeme', run_katello_installer=True, distribution=None):
+    """Task to install Nightly using forklift playbook"""
```
Comment (jhutar) [inline automation_tools/__init__.py]: Would it make sense to add description of the parameters into the docstring? Maybe at last the new one?

File: automation_tools/__init__.py [hunk 2]
```diff
@@ -1861,31 +1854,21 @@ def upstream_install(admin_password=None, run_katello_installer=True):
...
-    BASE_URL
-        URL for the compose repository.
-    SATELLITE_VERSION
-        Satellite version.
+def downstream_install(admin_password='changeme', run_katello_installer=True,
+                       sat_version='6.6', url=None):
+    """Task to install Satellite from internal repo
 
+    :param str sat_version: Indicates which Satellite version is being installed
+    :param str url: the repo URL where to install from
```
Comment (jhutar) [inline automation_tools/__init__.py]: New options are missing in the docstring.

File: automation_tools/__init__.py [hunk 3]
```diff
@@ -1910,28 +1893,21 @@ def downstream_install(admin_password=None, run_katello_installer=True):
...
-    The following environment variables affect this command:
-
-    ADMIN_PASSWORD
-        Optional, defaults to 'changeme'. Foreman admin password.
-    REPO_FILE_URL
-        URL for the compose repository file to fetch.
-    SATELLITE_VERSION
-        Satellite version.
+    :param str sat_version: Indicates which Satellite version is being installed
+    :param str url: the repofile URL where to install from
```
Comment (jhutar) [inline automation_tools/__init__.py]: Some options missing in the docstring?

File: automation_tools/__init__.py [hunk 4]
```diff
@@ -1956,23 +1932,15 @@ def repofile_install(admin_password=None, run_katello_installer=True,
...
-
-    The following environment variables affect this command:
+def ak_install(admin_password='changeme', run_katello_installer=True, sat_version='6.6'):
+    """Task to install Satellite via Activation Key from the Dogfood
 
-    ADMIN_PASSWORD
-        Optional, defaults to 'changeme'. Foreman admin password.
-    SATELLITE_VERSION
-        Satellite version.
+    :param str sat_version: Indicates which Satellite version is being installed
```
Comment (jhutar) [inline automation_tools/__init__.py]: Missing options in docstring here as well.

File: automation_tools/__init__.py [hunk 5]
```diff
@@ -1988,22 +1956,14 @@ def ak_install(admin_password=None, run_katello_installer=True):
...
-    The following environment variables affect this command:
-
-    ADMIN_PASSWORD
-        Optional, defaults to 'changeme'. Foreman admin password.
-    SATELLITE_VERSION
-        Satellite version.
+def cdn_install(admin_password='changeme', run_katello_installer=True, sat_version='6.5'):
+    """Task to install Satellite from CDN.
 
+    :param sat_version: Indicates which satellite version is being installed
```
Comment (jhutar) [inline automation_tools/__init__.py]: Same.

File: automation_tools/__init__.py [hunk 6]
```diff
@@ -2019,50 +1979,30 @@ def cdn_install(run_katello_installer=True):
...
-        The URL where the ISO will be downloaded.
-    ADMIN_PASSWORD
-        Optional, defaults to 'changeme'. Foreman admin password.
-    CHECK_GPG_SIGNATURES
-       Optional, all values other than 'true' will default to 'false'.
-    SATELLITE_VERSION
-        Satellite version.
+    :param bool check_gpg_signatures : Whether to check rpm gpg signatures or not
+    :param str sat_version: Indicates which Satellite version is being installed
+    :param str url: the ISO URL where to install from
```
Comment (jhutar) [inline automation_tools/__init__.py]: Same

File: automation_tools/__init__.py [hunk 7]
```diff
@@ -2121,13 +2060,17 @@ def product_install(distribution, create_vm=False, certificate_url=None,
     :param bool create_vm: creates a virtual machine and then install the
         product on it. Default: False.
     :param str certificate_url: where to fetch a fake certificate.
-    :param sat_version: Indicates which satellite version should be installed
-    :param puppet4: Indicates what puppet to install with Satellite 6.3
+    :param str sat_version: Indicates which satellite version should be installed
+    :param str puppet4: Indicates what puppet to install with Satellite 6.3
         Default: 'no', 'yes', 'upgrade'
```
Comment (jhutar) [inline automation_tools/__init__.py]: Again, some options are missing.

File: automation_tools/__init__.py [hunk 8]
```diff
@@ -2139,16 +2082,49 @@ def product_install(distribution, create_vm=False, certificate_url=None,
...
+        url = os.environ.get('BASE_URL')
+        if url:
+            task_params.update({'url': url})
+        else:
+            print('The BASE_URL environment variable should be defined')
+            sys.exit(1)
+    elif distribution == 'iso':
+        check_gpg_signatures = os.environ.get('CHECK_GPG_SIGNATURES', '').lower() == 'true'
+        task_params.update({'check_gpg_signatures': check_gpg_signatures})
+        url = os.environ.get('ISO_URL') or os.environ.get('BASE_URL')
```
Comment (jhutar) [inline automation_tools/__init__.py]: Why do we have `or` here? In the previous version it works only with `ISO_URL`. But, frankly, I have not investigated full logic here.

File: automation_tools/repository.py [hunk 9]
```diff
@@ -164,8 +164,7 @@ def enable_satellite_repos(cdn=False, beta=False, disable_enabled=True,
             repos.append('rhel-{0}-server-ansible-2.6-rpms')
             repos.append('rhel-{0}-server-satellite-maintenance-6-rpms')
         if sat_version == '6.6':
-            # Ansible '2.8' was not released yet, so using '2' repo
-            repos.append('rhel-{0}-server-ansible-2-rpms')
+            repos.append('rhel-{0}-server-ansible-2.8-rpms')
```
Comment (jhutar) [inline automation_tools/repository.py]: Thanks!

**Comment by san7ket** [comment]
I really like what you are doing here , I would also like to know the scope of the change needed further. 
We really need this, as we will be using groovy variables , and it doesn't make sense to again turn these groovy variables to ENV variables. 

Can you create a tracker for other changes that are needed to completely move away from env variables. 
This would greatly improve the traceability within the tasks and help quicker debugging.

File: automation_tools/__init__.py (hunk 8 above)
Comment (JacobCallahan) [reply automation_tools/__init__.py]: just reading it, that allows for it to use either field name. in the case an ISO_URL field isn't available/filled, then we can use the BASE_URL field.

File: automation_tools/__init__.py (hunk 2 above)
Comment (lpramuk) [reply automation_tools/__init__.py]: Old options are missing, gonna fill 'em up

File: automation_tools/__init__.py (hunk 1 above)
Comment (lpramuk) [reply automation_tools/__init__.py]: Ok, nice tip:-)

Comment (lpramuk) [reply automation_tools/__init__.py]: Added for ages omitted params ```admin_password``` and ```run_katello_installer``` into docstrings

File: automation_tools/__init__.py (hunk 8 above)
Comment (lpramuk) [reply automation_tools/__init__.py]: True.

File: automation_tools/repository.py (hunk 9 above)
Comment (lpramuk) [reply automation_tools/repository.py]: yw always :)

**Comment by lpramuk** [comment]
Added for ages omitted params ```admin_password``` and ```run_katello_installer``` into docstrings.

Do you guys now feel more ACKish? :-)

**Comment by san7ket** [comment]
@lpramuk I haven't reviewed it yet will get on it, but can you meanwhile also help track such occurrences in https://github.com/SatelliteQE/automation-tools/issues/799

**Comment by jyejare** [comment]
@lpramuk ping, is this PR valid and wanna work on it ? or shall we just review / merge it ?

**Comment by lpramuk** [comment]
@jyejare PR is valid and ready for review

File: automation_tools/__init__.py [hunk 10]
```diff
@@ -2088,36 +2083,22 @@ def product_install(distribution, certificate_url=None, selinux_mode=None, sat_v
...
-    if selinux_mode is None:
-        selinux_mode = os.environ.get('SELINUX_MODE', 'enforcing')
-
-    if distribution == 'satellite6-iso':
-        iso_url = os.environ.get('ISO_URL')
-        if iso_url is None:
-            print('The ISO_URL environment variable should be defined')
-            sys.exit(1)
+    if distribution == 'cdn' and sat_version not in ('6.3', '6.4', '6.5', '6.6', '6.7'):
+        raise ValueError("Satellite version should be in [6.3, 6.4, 6.5, 6.6, 6.7]")
```
Comment (JacobCallahan) [inline automation_tools/__init__.py:2087]: does 6.3 need to be removed from this section?

Comment (lpramuk) [reply automation_tools/__init__.py:2087]: it does need to be, but I am removing 6.3 functionality so it won't work properly anyway

Comment (lpramuk) [reply automation_tools/__init__.py:2087]: it doesn't need to be, I am removing 6.3 functionality in another PR https://github.com/SatelliteQE/automation-tools/pull/843 (merged)

So I better rebase to what's been merged

Comment (lpramuk) [reply automation_tools/__init__.py]: Rebased. (6.3 gone)

File: automation_tools/__init__.py [hunk 11]
```diff
@@ -993,8 +990,7 @@ def setup_foreman_discovery(sat_version):
         run(package_install('foreman-discovery-image'))
 
     # Unlock the default Locked template for discovery
-    run('hammer -u admin -p {0} template update '
-        '--name "PXELinux global default" --locked "false"'
+    run('hammer -u admin -p {0} template update --name "PXELinux global default" --locked "false"'
```
Comment (jyejare) [inline automation_tools/__init__.py:993]: Use Hammer function from Hammer module (https://github.com/SatelliteQE/automation-tools/blob/master/automation_tools/satellite6/hammer.py#L72) instead ?

File: automation_tools/__init__.py [hunk 12]
```diff
@@ -1781,38 +1781,28 @@ def upstream_install(admin_password=None, run_katello_installer=True, **kwargs):
...
 
     """
-    if admin_password is None:
-        admin_password = os.environ.get('ADMIN_PASSWORD', 'changeme')
-
-    base_url = os.environ.get('BASE_URL')
-    if base_url is None:
-        print('The BASE_URL environment variable should be defined')
+    if not url:
+        print('The url parameter should be defined')
```
Comment (jyejare) [inline automation_tools/__init__.py:1796]: Instead, raise some exception(probably ValueError here) so we need not to `sys.exit(1)` explicitly.

File: automation_tools/__init__.py [hunk 13]
```diff
@@ -2027,27 +1979,71 @@ def product_install(distribution, certificate_url=None, selinux_mode=None, sat_v
...
-    :param sat_version: Indicates which satellite version should be installed
+    :param str sat_version: Indicates which satellite version should be installed
 
     """
+    distribution = distribution.lower()
+    # Strip useless 'satellite6-' prefix if there is such
+    distribution = distribution[11:] if distribution.startswith('satellite6-') else distribution
     # Fetch the Satellite Version information.
     sat_version = sat_version or os.environ.get('SATELLITE_VERSION')
+    selinux_mode = selinux_mode or os.environ.get('SELINUX_MODE', 'enforcing')
```
Comment (jyejare) [inline automation_tools/__init__.py:1990]: Why this and below are not converted to params? At admin_passwd is converted everywhere in this conversion PR.

File: automation_tools/__init__.py [hunk 14]
```diff
@@ -2027,27 +1979,71 @@ def product_install(distribution, certificate_url=None, selinux_mode=None, sat_v
...
+        'run_katello_installer': False,
+        'admin_password': admin_password,
+    }
+    if distribution in ('upstream', 'koji'):
+        task_params.update({'distribution': distribution})
+    else:
+        task_params.update({'sat_version': sat_version})
+
+    if distribution == 'repofile':
+        url = os.environ.get('REPO_FILE_URL')
```
Comment (jyejare) [inline automation_tools/__init__.py:2019]: Why this is not converted ?

File: automation_tools/__init__.py (hunk 12 above)
Comment (jyejare) [reply automation_tools/__init__.py:1796]: Applicable everywhere in PR.

File: automation_tools/__init__.py (hunk 13 above)
Comment (jyejare) [reply automation_tools/__init__.py:1990]: Applicable for such env vars in PR. Why ?

File: automation_tools/__init__.py (hunk 11 above)
Comment (lpramuk) [reply automation_tools/__init__.py:993]: Out of scope of this PR, I only unwrap the line

File: automation_tools/__init__.py (hunk 13 above)
Comment (lpramuk) [reply automation_tools/__init__.py:1990]: The only place where I left ENV VARS is now `product_install()` - main task, from this task they are passed as params (and read from ENV VAR in subtask)

Tldr; this PR deals with using env vars in subtasks, for product_install I would raise another PR. Ok?

File: automation_tools/__init__.py (hunk 14 above)
Comment (lpramuk) [reply automation_tools/__init__.py:2019]: this PR deals with using env vars in subtasks, for product_install I would raise another PR. Ok?

I just dont like reading env vars on and on esp. in sub tasks when they should be passed as params

File: automation_tools/__init__.py (hunk 12 above)
Comment (lpramuk) [reply automation_tools/__init__.py]: Thanks.
Fixed. I also changed text of the messages from suggestion mode to error mode.

**Review by jyejare — APPROVED** [review]
ACKing based on promise to raise another PR for product_install task.
